### PR TITLE
IGNITE-14341 Improve pending tree cleanup

### PR DESCRIPTION
**Pull request by xtern**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 1]
```diff
@@ -6347,4 +6491,137 @@ protected String lockRetryErrorMessage(String op) {
...
+         * @param needOld {@code True} If need return old value.
+         */
+        protected RemoveRange(L lower, L upper, boolean needOld, int limit) {
+            super(lower, needOld, rmvRangeFromLeaf);
+
+            this.lower = lower;
+            this.upper = upper;
+
+            remaining = limit <= 0 ? -1 : limit;
+            removedRows = needOld ? new LinkedList<>() : null;
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:6531]: Not sure, but perhaps ArrayList will be more effective here (less GC pressure)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeSelfTest.java [hunk 2]
```diff
@@ -439,13 +465,13 @@ public void testPutRemove_1_20_pp_1() throws IgniteCheckedException {
      * @throws IgniteCheckedException If failed.
      */
     @Test
-    public void testPutRemove_1_20_pp_0() throws IgniteCheckedException {
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeSelfTest.java:442]: Why testPutRemove_1_20_pp_0 case was removed?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 3]
```diff
@@ -6347,4 +6491,137 @@ protected String lockRetryErrorMessage(String op) {
...
+            return true;
+        }
+
+        /** {@inheritDoc} */
+        @Override protected boolean ceil() {
+            return !completed;
+        }
+
+        /** {@inheritDoc} */
+        @Override protected void removeDataRowFromLeaf(long pageId, long page, long pageAddr, Boolean walPlc, BPlusIO<L> io,
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 4]
```diff
@@ -601,6 +602,104 @@ private class RemoveFromLeaf extends GetPageHandler<Remove> {
...
+            int rmvCnt = highIdx - idx + 1;
+
+            // Need to do inner replace when we remove the rightmost element and the leaf have no forward page,
+            // i.e. it is not the rightmost leaf of the tree.
+            boolean needReplaceInner = canGetRowFromInner && highIdx == cnt - 1 && io.getForward(leafAddr) != 0;
+
+            // !!! Before modifying state we have to make sure that we will not go for retry.
+
+            // We may need to replace inner key or want to merge this leaf with sibling after the remove -> keep lock.
+            if (needReplaceInner ||
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 1 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java]: Done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeSelfTest.java (hunk 2 above)
Comment (xtern) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeSelfTest.java]: Fixed, thanks

**Review by alex-plekhanov — APPROVED** [review]

**Pull request merged**